### PR TITLE
Mvfre 3/iteration

### PR DESCRIPTION
**Pull request by robtownsend1**

### CODE REVIEW

File: src/Contracts/ACTemporalExpression.php [hunk 1]
```diff
@@ -35,6 +51,68 @@ public function setFrequency(int $frequency): ACTemporalExpression
...
+     * @param DateTimeInterface $date
+     * @return bool
+     */
+    public function isIgnored(DateTimeInterface $date): bool
+    {
+        $ignoredDateStrings = array_map(function($d) {
+            return $d->toDateString();
+        }, $this->ignoreDates ? $this->ignoreDates : []);
+
+        return in_array($date->toDateString(), $ignoredDateStrings);
```
Comment (nickrupert7) [inline src/Contracts/ACTemporalExpression.php:74]: Super picky, but i actually see an opportunity for a performance improvement here. Technically with this code you're iterating over the `$this->ignoreDates` array twice - once to transform them all to strings, once to check for a match.

But there's a way where you should only have to do iterate once using a `for` loop

File: src/Contracts/ACTemporalExpression.php [hunk 2]
```diff
@@ -35,6 +51,68 @@ public function setFrequency(int $frequency): ACTemporalExpression
...
+     * Determine whether the current iteration date of the pattern is valid.
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        return $this->includes($this->current);
+    }
+
+    /**
+     * TODO: Implement for all child classes
```
Comment (nickrupert7) [inline src/Contracts/ACTemporalExpression.php:110]: You should be able to take this TODO comment out now

File: src/TemporalExpressions/TEDayOfMonth.php [hunk 3]
```diff
@@ -33,21 +45,27 @@ public static function build(DateTimeInterface $start, int $dayOfMonth): TEDayOf
...
-        $this->start = $start;
-        $this->dayOfMonth = $dayOfMonth;
+        $next = $this->current->addMonths($this->frequency);
+
+        if ($this->includes($next)) {
+            $this->current = $next;
+            return $next;
+        }
+
+        return null;
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDayOfMonth.php:59]: I think you're getting there, but there are a few problems. It looks like you didn't address the situation of seeking to a date which is not in the pattern, which will require a bunch of its own logic. As far as the logic for when `$this->current` IS in the pattern, I think you missed the case where you iterate to a date that was explicitly ignored. You'll want to have some logic that skips over those, and keeps pushing to the next instance until you run off the end of the pattern, or you find the next instance that is actually in the pattern.

File: src/TemporalExpressions/TEDayOfMonth.php [hunk 4]
```diff
@@ -33,21 +45,27 @@ public static function build(DateTimeInterface $start, int $dayOfMonth): TEDayOf
...
-     * @param DateTimeInterface $start Starting date of repetition pattern
-     * @param int $dayOfMonth Day of month (positive from beginning of month, negative from end of month)
+     * @inheritDoc
      */
-    public function __construct(DateTimeInterface $start, int $dayOfMonth)
+    public function next(): ?DateTimeInterface
     {
-        $this->start = $start;
-        $this->dayOfMonth = $dayOfMonth;
+        $next = $this->current->addMonths($this->frequency);
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDayOfMonth.php:52]: One thing about Carbon dates you have to be careful about is that the functions like `addMonths` actually modify the instance. So by doing this, you're actually modifying `$this->current` by adding months, and then creating a copy of it called `$next`, then your line below `$this->current = $next` actually doesn't do anything.

In moments like this you want to say `$next = $this->current->copy()->addMonths($this->frequency);`

File: src/TemporalExpressions/TEDayOfMonth.php [hunk 5]
```diff
@@ -33,21 +45,27 @@ public static function build(DateTimeInterface $start, int $dayOfMonth): TEDayOf
...
+        return null;
     }
 
     /**
      * @inheritDoc
      */
     public function includes(DateTimeInterface $date): bool
     {
+        //TODO: Add logic to check and ignore configured ignore dates
+
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDayOfMonth.php:68]: You should be able to take this out now

File: src/TemporalExpressions/TEDayOfWeekOfMonth.php [hunk 6]
```diff
@@ -37,16 +51,11 @@ public static function build(DateTimeInterface $start, int $dayOfWeek, int $week
...
-     * @param int $weekOfMonth Week of month (positive from beginning of month, negative from end of month)
+     * @inheritDoc
      */
-    public function __construct(DateTimeInterface $start, int $dayOfWeek, int $weekOfMonth)
+    public function next(): ?DateTimeInterface
     {
-        $this->start = $start;
-        $this->dayOfWeek = $dayOfWeek;
-        $this->weekOfMonth = $weekOfMonth;
+        // TODO: Implement next() method.
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDayOfWeekOfMonth.php:58]: I'm assuming this is one of the ones you intended to leave for me?

File: src/TemporalExpressions/TEDays.php [hunk 7]
```diff
@@ -29,12 +38,18 @@ public static function build(DateTimeInterface $start): TEDays
...
     {
-        $this->start = $start;
+        $next = $this->current->addDays($this->frequency);
+
+        if ($this->includes($next)) {
+            $this->current = $next;
+            return $next;
+        }
+
+        return null;
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDays.php:52]: Same

File: src/TemporalExpressions/TEDaysOfWeek.php [hunk 8]
```diff
@@ -34,16 +47,16 @@ public static function build(DateTimeInterface $start, $days): TEDaysOfWeek
...
     {
-        $this->validateIntArrayOrInt($days);
+        $next = $this->current->addWeeks($this->frequency);
 
-        $this->start = $start;
-        $this->days = is_array($days) ? $days : [$days];
+        if ($this->includes($next)) {
+            $this->current = $next;
+            return $next;
+        }
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDaysOfWeek.php:59]: This one actually isn't the same as the suggestion above - This one is a bit more complicated because it allows for you to select multiple weekdays right. So take a pattern that repeats every week on Monday and Friday - calling `next()` should return Monday, Friday, Monday, Friday, Monday, Friday, and so on. If you just add weeks, you'll just get Monday, Monday, Monday, over and over again, missing all of the Friday iterations.

Plus the fact that you have the option to set the frequency makes it a little trickier. If the pattern starts on a Monday , and repeats every two weeks on Monday and Friday, then it would be Monday, Friday 4 days later, Monday 10 days later, Friday 4 days later, etc. But if the pattern starts on a Friday, then it would be Friday, Monday 3 days later, Friday 11 days later, Monday 3 days later, etc.

In short, this pattern allows for a non-uniform incrementation pattern that we have to account for and it's not obviously as simple as jumping to the next weekday in the pattern either.

File: tests/TestCases/Contracts/TemporalExpressionTest.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        $endDate = new Carbon('2021-01-08');
+        $pattern = TEDays::build(new Carbon('2021-01-01'));
+        $nextDate;
+
+        if ($pattern->valid()) {
+            $nextDate = $pattern->next();
+        }
+
+        $this->assertEquals(
+            $nextDate ? $nextDate->toDateString() : '',
```
Comment (nickrupert7) [inline tests/TestCases/Contracts/TemporalExpressionTest.php:62]: You shouldn't need this ternary statement. In generic app code maybe, but in unit tests we should be controlling for every variable so there's no reason `$nextDate` should be `null` in this scenario

File: tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php [hunk 10]
```diff
@@ -135,4 +135,57 @@ public function testCorrectDateAcrossYearsWithCorrectFrequencyReturnsTrue()
 
         $this->assertTrue($result);
     }
+
+    public function testIgnoredDateInPatternReturnsFalse()
+    {
+        $pattern = TEDayofMonth::build(new Carbon('2021-01-01'), 1);
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php:141]: slight typo here `TEDayofMonth`

File: src/Contracts/ACTemporalExpression.php [hunk 11]
```diff
@@ -35,6 +51,68 @@ public function setFrequency(int $frequency): ACTemporalExpression
...
+    /**
+     * Set pattern iteration to the given date.
+     * @param DateTimeInterface $date
+     * @return DateTimeInterface
+     */
+    public function seek(DateTimeInterface $date): DateTimeInterface
+    {
+        if (static::includes($date)) {
+            $this->current = $date;
+        }
```
Comment (nickrupert7) [inline src/Contracts/ACTemporalExpression.php:96]: I like the idea of your approach here! Just soft-fail and don't accept a date if it's not valid.

I've been thinking about it more and more and I think in my mind it would make more sense if you could actually accept any date as the cursor, even if it's not in the pattern, then select the next valid date in the pattern from there. I know that makes things more complicated though, so we can discuss

File: tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php [hunk 12]
```diff
@@ -135,4 +135,57 @@ public function testCorrectDateAcrossYearsWithCorrectFrequencyReturnsTrue()
...
+            $next->toDateString(),
+            '2021-04-01'
+        );
+    }
+
+    public function testNextWithInvalidCurrentDateSelectsCorrectDate()
+    {
+        //TODO: Implement with seek() to invalid date, should select next valid date
+        $this->assertTrue(false);
+    }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php:178]: This is what i was talking about in the other comment about selecting _any_ date even if it's not in the pattern, then selecting the next valid date from there.

File: src/Contracts/ACTemporalExpression.php [hunk 13]
```diff
@@ -35,6 +51,68 @@ public function setFrequency(int $frequency): ACTemporalExpression
...
+    }
+
+    /**
+     * Determine whether the current iteration date of the pattern is valid.
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        return $this->includes($this->current);
+    }
```
Comment (nickrupert7) [inline src/Contracts/ACTemporalExpression.php:107]: It occurred to me that we may not actually want the criteria for `valid` to be that the current date is exactly in the pattern, but rather, that the current date is between the start and end date (or simply after the start date if no end date is set). The purpose of `valid` in the base PHP Iterator interface i showed you is to determine if you can keep iterating, but if we select a date that isn't in the pattern, but still within the valid range, we should be able to keep iterating.

File: src/TemporalExpressions/TEDayOfYear.php [hunk 14]
```diff
@@ -37,16 +51,18 @@ public static function build(DateTimeInterface $start, int $day, int $month): TE
...
-        $this->day = $day;
-        $this->month = $month;
+        $next = $this->current->addYears($this->frequency);
+
+        if ($this->includes($next)) {
+            $this->current = $next;
+            return $next;
+        }
+
+        return null;
```
Comment (nickrupert7) [inline src/TemporalExpressions/TEDayOfYear.php:65]: Same here

File: tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php [hunk 15]
```diff
@@ -135,4 +135,57 @@ public function testCorrectDateAcrossYearsWithCorrectFrequencyReturnsTrue()
...
+
+    public function testNextDateAfterPatternEndReturnsNull()
+    {
+        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1)
+            ->setEndDate(new Carbon('2021-06-01'));
+
+        while ($pattern->valid()) {
+            $pattern->next();
+        }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDayOfMonthTest.php:187]: I see a few problems here. Given the way you have `next()` implemented right now, and the only reason this doesn't result in an infinite loop is actually what I mentioned earlier about `addMonths()` actually modifying `$this->current` directly. Otherwise, because you never actually say `$this->current = $next` unless `$next` is in the pattern, you'd get an infinite loop stuck on the last date of the pattern.

In general I think unit tests should set the conditions a little more explicitly. When possible, avoid loops for situations because it obscures what's actually happening, and you can miss problems like the one I just pointed out. It would be better if you explicitly called `$pattern->next()` 6 times and checked the result after each one.

File: tests/TestCases/TemporalExpressions/TEDayOfYearTest.php [hunk 16]
```diff
@@ -111,4 +111,57 @@ public function testLeapDayMatchesMarch1OffLeapYearReturnsTrue()
...
+            $next->toDateString(),
+            '2030-01-01'
+        );
+    }
+
+    public function testNextWithInvalidCurrentDateSelectsCorrectDate()
+    {
+        //TODO: Implement with seek() to invalid date, should select next valid date
+        $this->assertTrue(false);
+    }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDayOfYearTest.php:154]: Same here, regarding seeking to a date not in the pattern

File: tests/TestCases/TemporalExpressions/TEDayOfYearTest.php [hunk 17]
```diff
@@ -111,4 +111,57 @@ public function testLeapDayMatchesMarch1OffLeapYearReturnsTrue()
...
+    {
+        $pattern = TEDayofYear::build(new Carbon('2021-01-01'), 1, 1)
+            ->setEndDate(new Carbon('2030-01-01'));
+
+        while ($pattern->valid()) {
+            $pattern->next();
+        }
+
+        $this->assertNull($pattern->next());
+    }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDayOfYearTest.php:166]: Same here, regarding the infinite loop, best practices in unit tests

File: tests/TestCases/TemporalExpressions/TEDaysTest.php [hunk 18]
```diff
@@ -70,4 +70,57 @@ public function testBasicCorrectDateReturnsTrue()
...
+            $next->toDateString(),
+            '2021-01-08'
+        );
+    }
+
+    public function testNextWithInvalidCurrentDateSelectsCorrectDate()
+    {
+        //TODO: Implement with seek() to invalid date, should select next valid date
+        $this->assertTrue(false);
+    }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDaysTest.php:113]: Same

File: tests/TestCases/TemporalExpressions/TEDaysTest.php [hunk 19]
```diff
@@ -70,4 +70,57 @@ public function testBasicCorrectDateReturnsTrue()
...
+    {
+        $pattern = TEDays::build(new Carbon('2021-01-01'))
+            ->setEndDate(new Carbon('2021-01-08'));
+
+        while ($pattern->valid()) {
+            $pattern->next();
+        }
+
+        $this->assertNull($pattern->next());
+    }
```
Comment (nickrupert7) [inline tests/TestCases/TemporalExpressions/TEDaysTest.php:125]: Same

**Review by nickrupert7 — CHANGES_REQUESTED** [review]
I had quite a bit of feedback on this one, but don't despair! I know the task I put to you here is pretty complicated and I'm seeing this as a learning opportunity with rounds of trial-and-error followed by feedback.

**Comment by nickrupert7** [comment]
> I had quite a bit of feedback on this one, but don't despair! I know the task I put to you here is pretty complicated and I'm seeing this as a learning opportunity with rounds of trial-and-error followed by feedback.

Also note since I'm taking over the task, you don't necessarily need to address all of my comments, because I very well may handle them before you get back from vacation.

**Comment by nickrupert7** [comment]
@robtownsend1 I would normally have left this open and worked with you through the feedback, but with you on vacation, I really wanted to get this one through, so I went ahead and opened my own PR to finish up the iteration logic. https://github.com/vetmoves/com.moves.php.fowler-recurring-events/pull/4.
